ls1intum/Artemis
# Complaints for Team Exercises

### Checklist

### Motivation and Context
Complaints and feedback requests have so far not been adapted for team-based exercises. This PR will allow teams to complain about the assessment of their submissions and to request more feedback. The complaint allowance in a course is separated for individual and team exercises. The instructor can set a max complaints limit per student that is valid for the individual exercises and a max complaints limit per team that is valid for the team exercises.

Example: A student in EIST has up to X rejected complaints for his individual work and additionally shares up to Y rejected complaints with his team for their team submissions. Feedback requests are generally not limited (as is the status quo for individual exercises).

The handling of complaints and more feedback requests is different for team-based exercises: Both complaints and more feedback requests are handled by the team tutor himself who is also the original assessor of the submission.

### Description
- Add field `maxTeamComplaints` to course and update course form (default is `3`, same as the default of `maxComplaints`)
- Add field `team` to complaint (and generalize `student`/`team` on complaint into `participant`)
- Check complaint allowance based on the exercise mode (limit for individual vs. team exercises)
- Adapt logic whether tutor is allowed to respond to complaints and feedback requests for the team exercise scenario (both client and server side)

### Steps for Testing
1. Log in to Artemis as a Student
2. Participate in a Team Exercise
3. Log in to Artemis as the Team Tutor
4. Assess the submission
5. Log in to Artemis as the Student
6. Complain about the assessment (variant: request more feedback)
7. Log in to Artemis as the Team Tutor
8. Respond to the complaint (variant: respond to the more feedback request)
9. Log in to Artemis as the Student
10. Check that the response to the complaint or more feedback request is shown

### Screenshots

# Student view
![OnPaste 20200530-153123](https://user-images.githubusercontent.com/7109225/83329761-94462400-a28b-11ea-95ab-f68f31d84475.png)
![OnPaste 20200530-153152](https://user-images.githubusercontent.com/7109225/83329763-95775100-a28b-11ea-914a-6a3905ba6a0d.png)
![OnPaste 20200530-153217](https://user-images.githubusercontent.com/7109225/83329764-95775100-a28b-11ea-8a82-703104b89654.png)


# Flow: Complaint
![OnPaste 20200530-153304](https://user-images.githubusercontent.com/7109225/83329765-960fe780-a28b-11ea-8a8a-998581c02036.png)
![OnPaste 20200530-153347](https://user-images.githubusercontent.com/7109225/83329766-960fe780-a28b-11ea-9a66-0ce34e7eb840.png)
![OnPaste 20200530-153446](https://user-images.githubusercontent.com/7109225/83329767-96a87e00-a28b-11ea-9dd6-632cf4f13bfc.png)
![OnPaste 20200530-153517](https://user-images.githubusercontent.com/7109225/83329768-96a87e00-a28b-11ea-93ec-166d348849af.png)


# Flow: More Feedback
![OnPaste 20200530-153611](https://user-images.githubusercontent.com/7109225/83329769-97411480-a28b-11ea-9d95-20c4e0baf185.png)
![OnPaste 20200530-153655](https://user-images.githubusercontent.com/7109225/83329770-97d9ab00-a28b-11ea-8abd-7cee22d96134.png)
![OnPaste 20200530-153715](https://user-images.githubusercontent.com/7109225/83329771-97d9ab00-a28b-11ea-90e6-2ca5a1855694.png)
![OnPaste 20200530-153754](https://user-images.githubusercontent.com/7109225/83329772-98724180-a28b-11ea-80b2-549d33f315d9.png)
